include/deal.II/lac/sparse_matrix_tools.h

```diff
@@ -262,22 +266,13 @@ namespace SparseMatrixTools
 
           const unsigned int row_length = sparsity_pattern.row_length(row);
 
-
-          std::pair<types::global_dof_index,
-                    std::vector<std::pair<types::global_dof_index, Number>>>
-            buffer;
           buffer.first = row;
+          buffer.second.resize(row_length);
 
-          for (unsigned int i = 0; i < row_length; ++i)
-            {
-              buffer.second.emplace_back(entry->column(), entry->value());
-
-              if (i + 1 != row_length)
-                ++entry;
-            }
+          for (unsigned int j = 0; j < row_length; ++j, ++entry)
+            buffer.second[j] = {entry->column(), entry->value()};
```

Comment:
The difference between the old and the new code is that `entry` was never incremented past the end of the row. I wonder whether that is for reasons such as those mentioned in #16414. I'm willing to see what happens here -- #16414 shows that we should fix the inability to increment past the end of the row, rather than having to deal with code such as what was here before.

Reply:
As far as I see it, the implementation is correct for incrementing out of bounds.